File: apps/desktop/crates/console-core/src/types/fs.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsEntry {
    pub name: String,
    pub path: String,
    /// Matches the server's `isDir` field (same as [`FsTreeEntry`]).
    pub is_dir: bool,
    pub size: Option<u64>,
    pub modified_at: Option<i64>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsTreeEntry {
    pub name: String,
    pub path: String,
    pub is_dir: bool,
    pub size: Option<u64>,
    pub git_status: Option<String>,
    pub children: Option<Vec<FsTreeEntry>>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BrowseDirectoryResponse {
    pub current_path: String,
    pub parent_path: Option<String>,
    pub entries: Vec<FsEntry>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DirectoryTreeResponse {
    pub path: String,
    pub tree_formatted: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileContentResponse {
    pub path: String,
    pub content: String,
    pub is_binary: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WriteFileDto {
    pub path: String,
    pub content: String,
}

/// Half-open byte range within [`GrepMatch::line_content`], used to bold the
/// matched substring in place (offsets, not rune/char indexes, matching the
/// server's fff-backed byte offsets).
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct GrepMatchRange {
    pub start: usize,
    pub end: usize,
}

/// One content-search hit from `GET /api/fs/grep`, shaped for the global
/// search panel: grouping by `rel_path`/`file_name`, highlighting via
/// `match_ranges`.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GrepMatch {
    pub rel_path: String,
    pub file_name: String,
    pub line_number: u64,
    pub column: i64,
    pub end_column: i64,
    pub line_content: String,
    pub match_ranges: Vec<GrepMatchRange>,
    #[serde(default)]
    pub is_binary: bool,
    #[serde(default)]
    pub is_definition: bool,
}

/// Response body of `GET /api/fs/grep`.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GrepResult {
    pub matches: Vec<GrepMatch>,
    pub total_matched: usize,
    pub files_searched: usize,
    pub total_files: usize,
    pub filtered_files: usize,
    pub next_cursor: u32,
    pub has_more: bool,
    /// Set when the query was an invalid regex and fff fell back to a
    /// literal search; results may still be usable.
    pub regex_error: Option<String>,
}

/// Content-search matching mode, mirrors the server's `fff.GrepMode`.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum GrepMode {
    Plain,
    #[default]
    Regex,
    Fuzzy,
}

impl GrepMode {
    pub fn as_query_value(self) -> &'static str {
        match self {
            Self::Plain => "plain",
            Self::Regex => "regex",
            Self::Fuzzy => "fuzzy",
        }
    }
}

/// Case-matching mode, mirrors the server's `fff.CaseMode`. Maps to the
/// search panel's "Aa" toggle (sensitive/insensitive) with smart as the
/// untoggled default.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum GrepCaseMode {
    #[default]
    Smart,
    Sensitive,
    Insensitive,
}

impl GrepCaseMode {
    pub fn as_query_value(self) -> &'static str {
        match self {
            Self::Smart => "smart",
            Self::Sensitive => "sensitive",
            Self::Insensitive => "insensitive",
        }
    }
}

/// Query parameters for [`crate::services::FsService::grep`], mirroring the
/// search panel's Aa / ab / .* toggles plus paging.
#[derive(Clone, Debug, Default)]
pub struct GrepOptions {
    pub mode: GrepMode,
    pub case_mode: GrepCaseMode,
    pub whole_word: bool,
    pub context_lines: Option<u32>,
    pub max_matches: Option<u32>,
    pub cursor: Option<u32>,
}

File: apps/server-go/internal/types/fs.go
// Filesystem types ported from packages/types/src/api.ts and fs.ts.
package types

type FsTreeEntry struct {
	Name      string        `json:"name"`
	Path      string        `json:"path"`
	IsDir     bool          `json:"isDir"`
	Size      *int64        `json:"size,omitempty"`
	GitStatus *string       `json:"gitStatus,omitempty"`
	Children  []FsTreeEntry `json:"children,omitempty"`
}

type FsBrowseResult struct {
	CurrentPath string        `json:"currentPath"`
	ParentPath  *string       `json:"parentPath"`
	Entries     []FsTreeEntry `json:"entries"`
}

// FileSearchResult mirrors the TS assist search item: the desktop ⌘P
// palette requires all four keys (score included).
type FileSearchResult struct {
	RelativePath string  `json:"relativePath"`
	AbsolutePath string  `json:"absolutePath"`
	IsDir        bool    `json:"isDir"`
	Score        float64 `json:"score"`
}

type FsChangeEvent struct {
	Type        string `json:"type"` // "fsChange"
	ProjectPath string `json:"projectPath"`
	EventPath   string `json:"eventPath,omitempty"`
}

type FilePreviewBlock struct {
	Kind    string `json:"kind"`
	Title   string `json:"title"`
	Message string `json:"message"`
}

// GrepMatchRange is a half-open byte range within GrepMatch.LineContent,
// used by the UI to bold the matched substring in place.
type GrepMatchRange struct {
	Start int `json:"start"`
	End   int `json:"end"`
}

// GrepMatch is the minimal content-search hit needed by the global search
// panel: identify and open the file, show its line, and highlight the match.
type GrepMatch struct {
	RelPath     string           `json:"relPath"`
	LineNumber  uint64           `json:"lineNumber"`
	LineContent string           `json:"lineContent"`
	MatchRanges []GrepMatchRange `json:"matchRanges"`
}

// GrepResult is the /api/fs/grep response payload. Scanner/index statistics
// stay inside the fff adapter and agent tool; the HTTP client only needs the
// displayed totals, page state, and any regex fallback warning.
type GrepResult struct {
	Matches       []GrepMatch `json:"matches"`
	TotalMatched  int         `json:"totalMatched"`
	FilteredFiles int         `json:"filteredFiles"`
	NextCursor    uint32      `json:"nextCursor"`
	HasMore       bool        `json:"hasMore"`
	RegexError    string      `json:"regexError,omitempty"`
}
